### PR TITLE
[MRG] Preassemble pipenv environments when no dependency on src

### PR DESCRIPTION
**Pull request by alex-treebeard**
As discussed in https://discourse.jupyter.org/t/repo2docker-image-caching/3292/2

We want to avoid pipenv installs when there has been no change in the lock file and the lock file has no dependency on the src directory.

Please point me in the direction of any code quality measures that are missing (I haven't written any tests but thought I'd get some feedback first)

Thanks!

### CODE REVIEW

**Comment by meeseeksmachine** [comment]
This pull request has been mentioned on **Jupyter Community Forum**. There might be relevant details there:

https://discourse.jupyter.org/t/repo2docker-image-caching/3292/3

**Comment by alex-treebeard** [comment]
Any chance I could get a review on here please? Would be great to know what the maintainers see as the remaining tasks before getting this merged in.

Looking forward to speeding up those pipenv builds.

**Comment by meeseeksmachine** [comment]
This pull request has been mentioned on **Jupyter Community Forum**. There might be relevant details there:

https://discourse.jupyter.org/t/treebeard-project-using-repo2docker/3600/4

**Comment by betatim** [comment]
Hi and sorry for the super slow review. You'd think having to stay home means lots of time for open-source, but somehow not :D 

I think one thing to add to this are some tests that automate the question "is this still working?" so that the feature continues to work. Pipfiles/envs are not something used by a lot of the maintainers of repo2docker so we need the automated checks to help keep it working. I have no idea if certain changes to this code would or wouldn't break things.

File: repo2docker/buildpacks/pipfile/__init__.py [hunk 1]
```diff
@@ -74,6 +75,43 @@ def python_version(self):
...
             return self._python_version
 
+    @property
+    def _should_preassemble_pipenv(self):
+        """Peek in Pipfile.lock to determine if we can assemble from only env files
+
+        If there are any local references, e.g. `"path": "./package.tar.gz"`,
+        stage the whole repo prior to installation.
+        """
+        if not os.path.exists("binder") and os.path.exists("setup.py"):
```
Comment (betatim) [inline repo2docker/buildpacks/pipfile/__init__.py]: I think this should use https://github.com/jupyter/repo2docker/blob/e6cbee56a8852674c817b1b2f17d040f4805f6ef/repo2docker/buildpacks/base.py#L503 to take care of `binder/` and `.binder/` (and what ever other sub-directories we might add in the future)

**Comment by alex-treebeard** [comment]
Great so integration tests cover both preassemble and non-preassemble use cases already for pipenv but I have added a unit tests file specifically to check the `_should_preassemble_pipenv` flag is working.

**Comment by alex-treebeard** [comment]
hi @betatim I think the only failing check was also failing in master, any chance we could trigger a rerun? (Sorry if I have missed an option for me to do this)

**Comment by alex-treebeard** [comment]
@consideRatio any chance you could take a look at this PR

I believe only failing tests are due to unrelated http timeouts
```
2020-04-05T11:39:52.7526885Z  ---> d400634b7be2
2020-04-05T11:39:52.7527321Z Step 2/8 : RUN pip install --no-cache notebook
2020-04-05T11:39:52.7527756Z  ---> Running in c6bad1ca4580
2020-04-05T11:39:52.7527966Z Collecting notebook
2020-04-05T11:39:52.7528387Z   Downloading notebook-6.0.3-py3-none-any.whl (9.7 MB)
2020-04-05T11:39:52.7528826Z [91mERROR: Exception:
2020-04-05T11:39:52.7529052Z Traceback (most recent call last):
2020-04-05T11:39:52.7529642Z   File "/usr/local/lib/python3.5/site-packages/pip/_vendor/urllib3/response.py", line 425, in _error_catcher
2020-04-05T11:39:52.7530154Z     yield
2020-04-05T11:39:52.7530661Z   File "/usr/local/lib/python3.5/site-packages/pip/_vendor/urllib3/response.py", line 507, in read
2020-04-05T11:39:52.7531034Z     data = self._fp.read(amt) if not fp_closed else b""
2020-04-05T11:39:52.7531371Z   File "/usr/local/lib/python3.5/http/client.py", line 458, in read
2020-04-05T11:39:52.7531736Z     n = self.readinto(b)
2020-04-05T11:39:52.7532011Z   File "/usr/local/lib/python3.5/http/client.py", line 498, in readinto
2020-04-05T11:39:52.7532307Z     n = self.fp.readinto(b)
2020-04-05T11:39:52.7532577Z   File "/usr/local/lib/python3.5/socket.py", line 576, in readinto
2020-04-05T11:39:52.7532855Z     return self._sock.recv_into(b)
2020-04-05T11:39:52.7533155Z   File "/usr/local/lib/python3.5/ssl.py", line 937, in recv_into
2020-04-05T11:39:52.7533435Z     return self.read(nbytes, buffer)
2020-04-05T11:39:52.7534567Z   File "/usr/local/lib/python3.5/ssl.py", line 799, in read
2020-04-05T11:39:52.7534896Z     return self._sslobj.read(len, buffer)
2020-04-05T11:39:52.7535178Z   File "/usr/local/lib/python3.5/ssl.py", line 583, in read
2020-04-05T11:39:52.7535453Z     v = self._sslobj.read(len, buffer)
2020-04-05T11:39:52.7535931Z ConnectionResetError: [Errno 104] Connection reset by peer
2020-04-05T11:39:52.7536149Z 
2020-04-05T11:39:52.7536563Z During handling of the above exception, another exception occurred:
2020-04-05T11:39:52.7536777Z 
2020-04-05T11:39:52.7536967Z Traceback (most recent call last):
2020-04-05T11:39:52.7537882Z   File "/usr/local/lib/python3.5/site-packages/pip/_internal/cli/base_command.py", line 186, in _main
2020-04-05T11:39:52.7538259Z     status = self.run(options, args)
2020-04-05T11:39:52.7538872Z   File "/usr/local/lib/python3.5/site-packages/pip/_internal/commands/install.py", line 331, in run
2020-04-05T11:39:52.7539381Z     resolver.resolve(requirement_set)
2020-04-05T11:39:52.7540005Z   File "/usr/local/lib/python3.5/site-packages/pip/_internal/legacy_resolve.py", line 177, in resolve
2020-04-05T11:39:52.7540455Z     discovered_reqs.extend(self._resolve_one(requirement_set, req))
2020-04-05T11:39:52.7541122Z   File "/usr/local/lib/python3.5/site-packages/pip/_internal/legacy_resolve.py", line 333, in _resolve_one
2020-04-05T11:39:52.7541551Z     abstract_dist = self._get_abstract_dist_for(req_to_install)
2020-04-05T11:39:52.7542422Z   File "/usr/local/lib/python3.5/site-packages/pip/_internal/legacy_resolve.py", line 282, in _get_abstract_dist_for
2020-04-05T11:39:52.7542854Z     abstract_dist = self.preparer.prepare_linked_requirement(req)
2020-04-05T11:39:52.7543564Z   File "/usr/local/lib/python3.5/site-packages/pip/_internal/operations/prepare.py", line 482, in prepare_linked_requirement
2020-04-05T11:39:52.7543929Z     hashes=hashes,
2020-04-05T11:39:52.7544501Z   File "/usr/local/lib/python3.5/site-packages/pip/_internal/operations/prepare.py", line 287, in unpack_url
2020-04-05T11:39:52.7544857Z     hashes=hashes,
2020-04-05T11:39:52.7545430Z   File "/usr/local/lib/python3.5/site-packages/pip/_internal/operations/prepare.py", line 159, in unpack_http_url
2020-04-05T11:39:52.7545819Z     link, downloader, temp_dir.path, hashes
2020-04-05T11:39:52.7546460Z   File "/usr/local/lib/python3.5/site-packages/pip/_internal/operations/prepare.py", line 303, in _download_http_url
2020-04-05T11:39:52.7546843Z     for chunk in download.chunks:
2020-04-05T11:39:52.7547392Z   File "/usr/local/lib/python3.5/site-packages/pip/_internal/utils/ui.py", line 160, in iter
2020-04-05T11:39:52.7547721Z     for x in it:
2020-04-05T11:39:52.7548278Z   File "/usr/local/lib/python3.5/site-packages/pip/_internal/network/utils.py", line 39, in response_chunks
2020-04-05T11:39:52.7548627Z     decode_content=False,
2020-04-05T11:39:52.7549197Z   File "/usr/local/lib/python3.5/site-packages/pip/_vendor/urllib3/response.py", line 564, in stream
2020-04-05T11:39:52.7549601Z     data = self.read(amt=amt, decode_content=decode_content)
2020-04-05T11:39:52.7550377Z   File "/usr/local/lib/python3.5/site-packages/pip/_vendor/urllib3/response.py", line 529, in read
2020-04-05T11:39:52.7550787Z     raise IncompleteRead(self._fp_bytes_read, self.length_remaining)
2020-04-05T11:39:52.7551128Z   File "/usr/local/lib/python3.5/contextlib.py", line 77, in __exit__
2020-04-05T11:39:52.7551425Z     self.gen.throw(type, value, traceback)
2020-04-05T11:39:52.7552134Z   File "/usr/local/lib/python3.5/site-packages/pip/_vendor/urllib3/response.py", line 443, in _error_catcher
2020-04-05T11:39:52.7552750Z     raise ProtocolError("Connection broken: %r" % e, e)
2020-04-05T11:39:52.7553947Z pip._vendor.urllib3.exceptions.ProtocolError: ("Connection broken: ConnectionResetError(104, 'Connection reset by peer')", ConnectionResetError(104, 'Connection reset by peer'))
```

**Comment by alex-treebeard** [comment]
@betatim any chance you can merge this please?

**Comment by consideRatio** [comment]
I looked through the code and I like your tests a lot, they were named and written in a way that helped me understand the PR itself better and that is a very good sign for future maintenance!

Seems like thorough nice work on this @alex-treebeard! :heart: :tada:

**Pull request closed without merging**